lib/grandpa/grandpa.go

```diff
@@ -449,188 +424,300 @@ func (s *Service) primaryBroadcastCommitMessage() {
 // playGrandpaRound executes a round of GRANDPA
 // at the end of this round, a block will be finalised.
 func (s *Service) playGrandpaRound() error {
-	logger.Debugf("starting round %d with set id %d",
-		s.state.round, s.state.setID)
+	currentRound := s.state.round
+	currentSetID := s.state.setID
+
 	start := time.Now()
 
-	ctx, cancel := context.WithCancel(s.ctx)
-	defer cancel()
+	logger.Debugf("starting round %d with set id %d",
+		currentRound, currentSetID)
 
-	isPrimary, err := s.handleIsPrimary()
-	if err != nil {
-		return err
-	}
+	cancel := make(chan struct{})
+	defer close(cancel)
 
-	logger.Debug("receiving pre-vote messages...")
-	go s.receiveVoteMessages(ctx)
-	time.Sleep(s.interval)
+	roundActions := s.finalizationEngine(cancel)
 
-	if s.paused.Load().(bool) {
-		return ErrServicePaused
-	}
+	for { //nolint:gosimple
+		select {
+		case action, ok := <-roundActions:
+			if !ok {
+				return nil
+			}
 
-	// broadcast pre-vote
-	pv, err := s.determinePreVote()
-	if err != nil {
-		return err
+			switch action {
+			case determinePrevote:
+				isPrimary, err := s.handleIsPrimary()
+				if err != nil {
+					return err
+				}
+
+				// broadcast pre-vote
+				preVote, err := s.determinePreVote()
+				if err != nil {
+					return err
+				}
+
+				signedpreVote, prevoteMessage, err :=
+					s.createSignedVoteAndVoteMessage(preVote, prevote)
+				if err != nil {
+					return err
+				}
+
+				if !isPrimary {
+					s.prevotes.Store(s.publicKeyBytes(), signedpreVote)
+				}
+
+				s.sendPrevoteMessage(prevoteMessage)
+
+			case determinePrecommit:
+				preComit, err := s.determinePreCommit()
+				if err != nil {
+					return fmt.Errorf("determining precommit: %w", err)
+				}
+
+				signedpreComit, precommitMessage, err :=
+					s.createSignedVoteAndVoteMessage(preComit, precommit)
+				if err != nil {
+					return fmt.Errorf("creating signed vote: %w", err)
+				}
+
+				s.precommits.Store(s.publicKeyBytes(), signedpreComit)
+
+				s.sendPrecommitMessage(precommitMessage)
+
+			case finalize:
+				commitMessage, err := s.newCommitMessage(s.head, s.state.round, s.state.setID)
+				if err != nil {
+					return fmt.Errorf("generating commit message: %w", err)
```

Comment:
nit creating feels more appropriate than `generating` (no randomness etc. if you see what I mean)

```suggestion
					return fmt.Errorf("creating commit message: %w", err)
```